src/dune_pkg/local_package.ml

```diff
@@ -108,16 +64,82 @@ module For_solver = struct
          (Package_dependency.list_to_opam_filtered_formula depopts)
   ;;
 
-  let opam_filtered_dependency_formula { dependencies; _ } =
-    Package_dependency.list_to_opam_filtered_formula dependencies
+  let opam_filtered_dependency_formula { dependencies; _ } = dependencies
+
+  module T = struct
+    type t = Package_name.t * OpamTypes.condition
+
+    let to_dyn = Dyn.pair Package_name.to_dyn Opam_dyn.condition
+
+    let compare (name, condition) (other_name, other_condition) =
+      let open Ordering.O in
+      let= () = Package_name.compare name other_name in
+      let compare_filters filter filter' =
+        let get_vars = function
+          | OpamTypes.Constraint _ -> []
+          | Filter filter -> Stdlib.List.sort Stdlib.compare (OpamFilter.variables filter)
+        in
+        match get_vars filter, get_vars filter' with
+        | v :: _, v' :: _ -> Stdlib.compare v v'
+        | [], _ :: _ -> 1
+        | _ :: _, [] -> -1
+        | [], [] -> 0
+      in
+      Ordering.of_int
+        (OpamFormula.compare_formula compare_filters condition other_condition)
+    ;;
+  end
+
+  module Map = Map.Make (T)
+  module Set = Set.Make (T) (Map)
+
+  let maximum_package_set formula =
```

Comment:
That is true, but bar of evaluating the formula with all the filters (which would need access to all of the OPAM variables and lead to potentially different solutions on different systems) this is a simple way to determine an overapproximation of the non-local packages. In the worst case this would lead to

1. Reporting the non-local package error with non-local package that wouldn't get picked (as we only ever pick one at random for the error message)
2. The hash of non-local packages would change if a non-local package changes its dependencies even if we don't actually end up depending on it.

While only an approximation I think the upside is a stable solution across systems. The solver still gets the whole dependency formula and will compute an accurate lock file.

Reply:
Btw, any folds in the opam library you could reach for to write this?

Reply:
Good point. Yes, ended up doing exactly that. Also given `"foo" | "bar"` and `"bar" | "foo"` lead to different solutions.